File: src/main/kotlin/io/titandata/titan/providers/local/Log.kt
/*
 * Copyright (c) 2019 by Delphix. All rights reserved.
 */

package io.titandata.titan.providers.local

import io.titandata.client.apis.CommitsApi

class Log (
    private val commitsApi: CommitsApi = CommitsApi()
) {
    private val n = System.lineSeparator()

    fun log(container: String, tags: List<String>) {
        var first = true
        for (commit in commitsApi.listCommits(container, tags)) {
            if (!first) {
                println("")
            } else {
                first = false
            }
            val metadata = commit.properties
            println("commit ${commit.id}")
            if (metadata.containsKey("author")) {
                println("Author: ${metadata["author"]}")
            }
            if (metadata.containsKey("user")) {
                println("User: ${metadata["user"]}")
            }
            if (metadata.containsKey("email")) {
                println("Email: ${metadata["email"]}")
            }
            println("Date: ${metadata["timestamp"]}")
            if (metadata.containsKey("tags")) {
                @Suppress("UNCHECKED_CAST")
                val tags = metadata.get("tags") as Map<String, String>
                if (!tags.isEmpty()) {
                    print("Tags:")
                    for ((key, value) in tags) {
                        print(" ")
                        if (value != "") {
                            print("$key=$value")
                        } else {
                            print(key)
                        }
                    }
                    println("")
                }
            }
            if (metadata["message"] != "") {
                println("${n}${metadata["message"]}")
            }
        }
    }
}

File: internal/app/providers/common/Log.go
package common

import (
	"fmt"
	"github.com/antihax/optional"
	client "github.com/titan-data/titan-client-go"
	"strconv"
)

func Log(repo string, tags []string, port int) {
	cfg.BasePath = "http://localhost:" + strconv.Itoa(port)

	first := true
	o := optional.NewInterface(tags)
	opts := client.ListCommitsOpts{Tag:o}
	commits, _, _ := commitsApi.ListCommits(ctx, repo, &opts)

	for _, commit := range commits {
		if !first {
			fmt.Println("")
		} else {
			first = false
		}
		metadata := commit.Properties
		fmt.Println("commit " + commit.Id)
		ifContainsPrint(metadata, "author")
		ifContainsPrint(metadata, "user")
		ifContainsPrint(metadata, "email")
		ifContainsPrint(metadata, "timestamp")

		tags, ok := metadata["tags"].(map[string]interface{})
		if ok {
			fmt.Print("Tags: ")
			for t, v := range tags {
				if len(v.(string)) > 0 {
					fmt.Printf("%v=%v ", t, v)
				} else  {
					fmt.Printf("%v ", t)
				}
			}
			fmt.Println()
		}

		if metadata["message"] != "" {
			out := fmt.Sprintf("\n%v", metadata["message"])
			fmt.Println(out)
		}
	}
}


